eras/conway/impl/src/Cardano/Ledger/Conway/Governance/DRepPulser.hs

```diff
@@ -160,60 +159,39 @@ instance EraPParams era => ToCBOR (PulsingSnapshot era) where
+-- that Credential to the DRep Distribution, if the DRep is a DRepCredential
+-- (rather than AlwaysAbstain or AlwaysNoConfidence) and a member of the
+-- registered DReps.
 --
---  Compute the Drep distribution of stake(Coin)
---  cost is expected to be O(size of 'stakeDistr' * log (size of 'um') * log (size of 'regDreps'))
---  This is going to be expensive, so we will want to pulse it. Without pulsing,
---  we estimate 3-5 seconds
-computeDrepDistr ::
+-- TODO: precisely workout the complexity and add to this documentation.
+computeDRepDistr ::
+  forall c.
   UMap c ->
   Map (Credential 'DRepRole c) (DRepState c) ->
   Map (Credential 'Staking c) (CompactForm Coin) ->
-  Map (DRep c) (CompactForm Coin)
-computeDrepDistr um regDreps stakeDistr =
-  Map.foldlWithKey' (accumDRepDistr um regDreps) Map.empty stakeDistr
-
--- | For each 'stakecred' and coin 'c', check if that credential is delegated to some DRep.
---   If so then add that coin to the aggregated map 'ans', mapping DReps to compact Coin
---   If the DRep is a DRepCredential (rather than AwaysAbstain or AlwaysNoConfidence) then check
---   that the credential is a member of the registered DRep map ('regDreps') before adding it to 'ans'
-accumDRepDistr ::
-  UMap c ->
-  Map (Credential 'DRepRole c) (DRepState c) ->
   Map (DRep c) (CompactForm Coin) ->
-  Credential 'Staking c ->
-  CompactForm Coin ->
   Map (DRep c) (CompactForm Coin)
-accumDRepDistr um regDreps ans stakeCred stake@(CompactCoin compactStake) = fromMaybe ans $ do
-  umElem <- Map.lookup stakeCred (umElems um)
-  drep <- umElemDRep umElem
-  let stakeWithRewards =
-        case umElemRDPair umElem of
-          Nothing -> stake
-          Just rdPair
-            | CompactCoin compactReward <- rdReward rdPair ->
-                CompactCoin (compactReward + compactStake)
-  pure $ case drep of
-    DRepAlwaysAbstain -> Map.insertWith UMap.addCompact drep stakeWithRewards ans
-    DRepAlwaysNoConfidence -> Map.insertWith UMap.addCompact drep stakeWithRewards ans
-    DRepCredential drepCred
-      | Map.member drepCred regDreps -> Map.insertWith UMap.addCompact drep stakeWithRewards ans
-      | otherwise -> ans
+computeDRepDistr um regDReps stakeDistr dRepDistr =
+  let stakesAndRewards =
+        Map.unionWith
+          addCompact
+          stakeDistr
+          $ fmap rdReward
+          $ Map.mapMaybe umElemRDDRepActive
+          $ umElems um
+      dreps = dRepMap um
+      accum distr stakeCred drep =
+        let stakeAndReward = fromMaybe (CompactCoin 0) $ Map.lookup stakeCred stakesAndRewards
+            distrWithStakeAndReward = Map.insertWith addCompact drep stakeAndReward distr
+         in case drep of
+              DRepAlwaysAbstain -> distrWithStakeAndReward
+              DRepAlwaysNoConfidence -> distrWithStakeAndReward
+              DRepCredential drepCred
+                | Map.member drepCred regDReps -> distrWithStakeAndReward
+                | otherwise -> dRepDistr
```

Comment:
I wonder if this shouldn't be `distr`  instead of `dRepDistr`, since it's the accumulator value within the `accum` function, pertaining to it's sibling `drep` argument.

Reply:
Awesome catch! Thank you! :raised_hands: